File: app/src/main/java/com/forkmang/adapter/Pickup_Fragment_BookTableAdapter.java
package com.forkmang.adapter;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RatingBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.forkmang.R;
import com.forkmang.activity.PickupSelectFood_Activity;
import com.forkmang.data.RestoData;

import java.util.ArrayList;


public class Pickup_Fragment_BookTableAdapter extends RecyclerView.Adapter<Pickup_Fragment_BookTableAdapter.BookTableItemHolder> {
    Activity activity;
    ArrayList<RestoData> resto_dataArrayList;

    public Pickup_Fragment_BookTableAdapter(Activity activity, ArrayList<RestoData> resto_dataArrayList, String coming_from, Context ctx) {
        this.activity = activity;
        this.resto_dataArrayList = resto_dataArrayList;
    }

    public Pickup_Fragment_BookTableAdapter(Activity activity) {
        this.activity = activity;
    }

    /*book_table_cell*/

    @NonNull
    @Override
    public BookTableItemHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.book_table_cell, null);
        BookTableItemHolder bookTableItemHolder = new BookTableItemHolder(v);
        return bookTableItemHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull BookTableItemHolder holder, int position) {
        RestoData bookTable = resto_dataArrayList.get(position);
        holder.txtrestroname.setText(bookTable.getRest_name());
        holder.txt_endtime.setText(bookTable.getEndtime());
        holder.txttotalkm.setText(bookTable.getDistance()+" Km");
    }

    @Override
    public int getItemCount(){
        return resto_dataArrayList.size();
    }

    public class BookTableItemHolder extends RecyclerView.ViewHolder {

        ImageView imgproduct;
        TextView txtrestroname, txt_endtime, txt_distance,txttotalkm,txt_ratingno;
        RatingBar rating_bar;
        RelativeLayout relative_view;

        public BookTableItemHolder(@NonNull View itemView) {
            super(itemView);
            relative_view = itemView.findViewById(R.id.relative_view);
            txtrestroname=itemView.findViewById(R.id.txtrestroname);
            txt_endtime=itemView.findViewById(R.id.txt_endtime);
            txt_distance=itemView.findViewById(R.id.txt_distance);
            txttotalkm=itemView.findViewById(R.id.txt_totalkm);
            txt_ratingno=itemView.findViewById(R.id.txt_ratingno);
            imgproduct=itemView.findViewById(R.id.imgrestro);
            rating_bar=itemView.findViewById(R.id.rating_bar);

            relative_view.setOnClickListener(v -> {
                int position = getBindingAdapterPosition();
                RestoData restoData = resto_dataArrayList.get(position);
                final Intent mainIntent = new Intent(activity, PickupSelectFood_Activity.class);
                mainIntent.putExtra("restromodel", restoData);
                activity.startActivity(mainIntent);

                //activity.finish();
            });

        }
    }

}

File: app/src/main/java/com/forkmang/adapter/Pickup_Fragment_BookTableAdapter.kt
package com.forkmang.adapter

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RatingBar
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.forkmang.R
import com.forkmang.data.RestoData

class Pickup_Fragment_BookTableAdapter(
    var onItemClicked: ((clickPosition: RestoData) -> Unit)
) : ListAdapter<RestoData, Pickup_Fragment_BookTableAdapter.BookTableItemHolder>(ItemCallback) {

    object ItemCallback : DiffUtil.ItemCallback<RestoData>() {
        override fun areItemsTheSame(oldItem: RestoData, newItem: RestoData): Boolean =
            oldItem.id == newItem.id

        @SuppressLint("DiffUtilEquals")
        override fun areContentsTheSame(oldItem: RestoData, newItem: RestoData): Boolean =
            oldItem == newItem
    }

    var resto_dataArrayList: List<RestoData> = emptyList()
        set(value) {
            field = value
            submitList(field)
        }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): BookTableItemHolder {
        val v: View =
            LayoutInflater.from(parent.context).inflate(R.layout.book_table_cell, null)
        return BookTableItemHolder(v)
    }


    override fun onBindViewHolder(holder: BookTableItemHolder, position: Int) {
        val bookTable: RestoData = resto_dataArrayList[position]
        holder.txtrestroname.text = bookTable.rest_name
        holder.txt_endtime.text = bookTable.endtime
        holder.txttotalkm.text = bookTable.distance + " Km"
    }


    inner class BookTableItemHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var imgproduct: ImageView
        var txtrestroname: TextView
        var txt_endtime: TextView
        var txt_distance: TextView
        var txttotalkm: TextView
        var txt_ratingno: TextView
        var rating_bar: RatingBar
        var relative_view: RelativeLayout

        init {
            relative_view = itemView.findViewById<RelativeLayout>(R.id.relative_view)
            txtrestroname = itemView.findViewById<TextView>(R.id.txtrestroname)
            txt_endtime = itemView.findViewById<TextView>(R.id.txt_endtime)
            txt_distance = itemView.findViewById<TextView>(R.id.txt_distance)
            txttotalkm = itemView.findViewById<TextView>(R.id.txt_totalkm)
            txt_ratingno = itemView.findViewById<TextView>(R.id.txt_ratingno)
            imgproduct = itemView.findViewById<ImageView>(R.id.imgrestro)
            rating_bar = itemView.findViewById<RatingBar>(R.id.rating_bar)
            relative_view.setOnClickListener {
                val position: Int = bindingAdapterPosition
                val restoData: RestoData = resto_dataArrayList[position]
                onItemClicked(restoData)
            }
        }
    }
}